Feature/hibernate

S1mm1ssim opened the pull request:
1. Migrate from JDBC to Hibernate:
- Add hibernate dependencies
- Update DAO layer to use Hibernate
- Add mapping on Entity classes
- Add optimistic locking mechanism on Item and Position entity classes
2. Implement pagination and filtration;
3. Add test dependencies
4. Implement unit tests for service classes and for custom CategoryDto to Category mapper
5. Setup postgres test container for integration and dao tests
6. Setup hibernate session factory for tests

S1mm1ssim committed:
Configure hibernate

S1mm1ssim committed:
Implement entity classes and update migrations

- Annotate entity classes
- Implement their EntityGraphs
- Create migrations, add version column to implement optimistic locks on Item and Position
- Add check constraint on Position.amount

S1mm1ssim committed:
Remove "created" and "updated" field from DTO classes and add "version" for ItemDto and PositionDto

S1mm1ssim committed:
Add pageNumber param for getAll() and instantiate LocalDateTime.now() values for "created" and "updated" entity fields

S1mm1ssim committed:
Implement global REST exception handler

S1mm1ssim committed:
Add missing logs

S1mm1ssim committed:
Rename entity graphs and their subgraphs

S1mm1ssim committed:
Fix logs

S1mm1ssim committed:
Update Dao interface to support pagination and filtering

S1mm1ssim committed:
Update Item, Category and Position services' getAll method to use emptyMap() as they do not provide filtration

S1mm1ssim committed:
Add utility methods

S1mm1ssim committed:
Implement Company pagination and filtration in service and controller

S1mm1ssim committed:
Implement User pagination and filtration in service and controller

S1mm1ssim committed:
Implement UserDao.java with pagination and filtration

S1mm1ssim committed:
Implement Company.java with pagination and filtration

S1mm1ssim committed:
Remove unused import

S1mm1ssim committed:
Implement CategoryDao.java

S1mm1ssim committed:
Implement ItemDao.java

S1mm1ssim committed:
Implement PositionDao.java

S1mm1ssim committed:
Add test maven dependencies

S1mm1ssim committed:
Implement optimistic locking mechanism for Item and Position entities

S1mm1ssim committed:
Remove unused EntityGraph

S1mm1ssim committed:
Update logs to print requested page number

S1mm1ssim committed:
Move default user role to properties file and rename file.size to default.file.size

S1mm1ssim committed:
Refactor SessionFactoryConfig. Create bean of type SessionFactory. Update DI in DAO classes

S1mm1ssim committed:
Remove unused constants

S1mm1ssim committed:
Remove sessionFactory.getSessionFactory() call

S1mm1ssim committed:
Update test dependencies and configure maven.surefire plugin

S1mm1ssim committed:
Change hbm2ddl.auto to none

S1mm1ssim committed:
Implement service unit tests

S1mm1ssim committed:
Testcontainers setup for DAO layer testing

S1mm1ssim committed:
Fix testcontainers startup

S1mm1ssim committed:
Refactor: implement hibernate configuration with java code and use properties from application.yml file

S1mm1ssim committed:
Update rest assured dependency

S1mm1ssim committed:
Implement Testcontainers setup:

- single container init for all tests
- on init dynamically set connection properties for Hibernate SessionFactory

S1mm1ssim committed:
Implement user controller integration tests

S1mm1ssim committed:
Implement CategoryDao unit test

S1mm1ssim committed:
Bugfix: fix mapper to correctly handle all input value combinations for parent category mapping and enable corresponding test

S1mm1ssim committed:
Update javadoc

S1mm1ssim committed:
Update several tests to correspond to new contract on CategoryDto to Category mapping

S1mm1ssim committed:
Bugfix: when executing update with all update properties being null update exception was thrown.

Now no query is attempted to be executed if no update values are provided.

S1mm1ssim committed:
Fix CategoryMapper throwing NullPointerException in some cases

S1mm1ssim committed:
Remove unnecessary setUp method

S1mm1ssim committed:
Implement CompanyDao unit tests

S1mm1ssim committed:
Bugfix: don't set parent entity's id but rather set child's parent entity with new object value

S1mm1ssim committed:
Implement ItemDao unit tests

S1mm1ssim committed:
Bugfix: fix 500 status if user provided non-existent companyId for update operation

S1mm1ssim committed:
Implement UserDao unit tests

S1mm1ssim committed:
Change test logging level to debug

S1mm1ssim committed:
Remove code duplication by calling superclass method

S1mm1ssim committed:
Rename sql file that rolls database back before tests or their data.sql file were executed

S1mm1ssim committed:
Fix logs

S1mm1ssim committed:
Implement PositionController integration tests

S1mm1ssim committed:
Update postman collections

Versach commented on src/main/java/com/modsensoftware/marketplace/config/SessionFactoryConfig.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Bean
+    @DependsOn("hibernateProperties")
+    public SessionFactory sessionFactory() {
```
какой то сложный путь конфигурации, можно сделать проще, зачем именно так явно сделал?

S1mm1ssim replied:
Да, DependsOn лишний, согласен. Всё исправлено

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/ItemController.java:
```diff
@@ -30,18 +31,21 @@ public class ItemController {
+            defaultValue = DEFAULT_PAGE_NUMBER) int pageNumber) {
         if (log.isDebugEnabled()) {
-            log.debug("Fetching all users");
+            log.debug("Fetching all items");
```
предлагаю написать для целей дебага аспект

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/PositionController.java:
```diff
@@ -29,12 +30,15 @@ public class PositionController {
 
     private final PositionService positionService;
 
+    private static final String DEFAULT_PAGE_NUMBER = "0";
```
может стоит вынести это в отдельный файл с константами или в интерфейст, если есть еще что то общее между контроллерами? как то неочень повторять очевидно стандартную для всех pageable контроллеров константу

S1mm1ssim replied:
Вынес в класс с константами

Versach commented on src/main/java/com/modsensoftware/marketplace/dao/CategoryDao.java:
```diff
@@ -23,161 +29,124 @@
+        CriteriaQuery<Category> byId = cb.createQuery(Category.class);
+        Root<Category> root = byId.from(Category.class);
+        Join<Category, Category> parentCategory =
+                root.join(CATEGORY_PARENT_FIELD_NAME, JoinType.LEFT);
```
мне кажется в таком случае константам *_field_name тогда уж место в энтити

S1mm1ssim replied:
Вынес в энтити

Versach commented on src/main/java/com/modsensoftware/marketplace/service/impl/CompanyServiceImpl.java:
```diff
@@ -28,23 +34,37 @@ public Company getCompanyById(Long id) {
         }
-        return companyDao.getAll();
+        Map<String, String> filterProperties = new HashMap<>();
+        Utils.putIfNotNull("email", email, filterProperties::put);
```
прям очень некрасиво константы строковые тут

S1mm1ssim replied:
Вынес в класс с константами

Versach commented on src/main/java/com/modsensoftware/marketplace/service/impl/ItemServiceImpl.java:
```diff
@@ -61,6 +68,15 @@ public void updateItem(UUID id, ItemDto updatedFields) {
+        } else {
+            log.error("Item versions do not match. Provided: {}, in the database: {}",
+                    updatedFields.getVersion(), item.getVersion());
+            throw new OptimisticLockException("Provided item version "
```
когда мы говорим о подобных сообщегиях их стоит выность в константы а лучше в проперти - упрощает дальнешую локализацию и в целом разделяет "код" и клиенские сообщения грубо говоря

S1mm1ssim replied:
Все exception сообщения вынес в проперти

Versach commented on src/main/java/com/modsensoftware/marketplace/service/impl/PositionServiceImpl.java:
```diff
@@ -28,23 +30,28 @@ public Position getPositionById(Long id) {
-        positionDao.save(positionMapper.toPosition(positionDto));
+        Position position = positionMapper.toPosition(positionDto);
+        position.setCreated(LocalDateTime.now());
+        if (log.isDebugEnabled()) {
```
А нельзя это просто уровнем дебага задавать вместо того чтобы ифами обвешивать? зачем нам вообще тогда уровни логирования

S1mm1ssim replied:
Исправил, убрал везде эти if (log.isDebugEnabled())

Versach commented on src/main/java/com/modsensoftware/marketplace/service/impl/UserServiceImpl.java:
```diff
@@ -24,28 +32,56 @@ public class UserServiceImpl implements UserService {
+        Utils.putIfNotNull("email", email, filterProperties::put);
+        Utils.putIfNotNull("name", name, filterProperties::put);
+        if (createdBetween != null) {
+            filterProperties.put("created", createdBetween);
```
все в константы, вот этих назвний полей и прочего тут быть не должно

S1mm1ssim replied:
Вынес в класс с константами

Versach commented on src/test/java/com/modsensoftware/marketplace/integration/AbstractIntegrationTest.java:
```diff
@@ -0,0 +1,44 @@
+    protected static void beforeAll() {
+        String port = System.getProperty("server.port");
+        if (port == null) {
+            RestAssured.port = 8081;
```
в тестах допустимо но опять же - лучше в константу какую нибудь вынести порт

S1mm1ssim replied:
Готово

Versach commented on src/test/java/com/modsensoftware/marketplace/integration/position/PositionControllerIntegrationTest.java:
```diff
@@ -0,0 +1,161 @@
+                .then().statusCode(404);
+    }
+
+    @Test
+    public void shouldReturn404StatusIfPositionsCompanyIsSoftDeleted() {
+        String positionId = "1001";
+        RestAssured.given()
+                .contentType("application/json")
+                .when()
+                .get(format("/positions/%s", positionId))
+                .then().statusCode(404);
+    }
+
+    @Test
+    public void shouldReturn204StatusOnDeleteOperation() {
+        String positionId = "1002";
+        RestAssured.given()
+                .contentType("application/json")
+                .when()
+                .delete(format("/positions/%s", positionId))
+                .then().statusCode(204);
+    }
```
Можно объеденить в parametrized test особенно если их будет больше

S1mm1ssim replied:
Готово

Versach commented on src/test/java/com/modsensoftware/marketplace/unit/company/CompanyDaoTest.java:
```diff
@@ -0,0 +1,331 @@
+    }
+
+    @Test
+    public void existsByEmailShouldBeFalseIfSoftDeleted() {
+        // given
+        Company company = new Company(null, "name", "email@email.com",
+                LocalDateTime.now().truncatedTo(ChronoUnit.SECONDS), "description", true);
+        underTest.save(company);
+        // when
+        boolean existsByEmail = underTest.existsByEmail(company.getEmail());
+        // then
+        Assertions.assertThat(existsByEmail).isFalse();
+        // clean up
+        Session session = sessionFactory.openSession();
+        Transaction transaction = session.beginTransaction();
+        session.delete(company);
+        transaction.commit();
+        session.close();
+    }
+
+    @Test
+    public void existsByEmailShouldBeFalseIfEntityNotFound() {
+        // given
+        Company company = new Company(null, "name", "email@email.com",
+                LocalDateTime.now().truncatedTo(ChronoUnit.SECONDS), "description", false);
+        underTest.save(company);
+        // when
+        boolean existsByEmail = underTest.existsByEmail(company.getEmail() + "dsahjshfjdskfd");
+        // then
+        Assertions.assertThat(existsByEmail).isFalse();
+        // clean up
+        Session session = sessionFactory.openSession();
+        Transaction transaction = session.beginTransaction();
+        session.delete(company);
+        transaction.commit();
+        session.close();
+    }
+
```
Тоже напрашивается parameterized Test мб еще где то, вникни сам где разумнее их использовать

S1mm1ssim replied:
Этот сделал и нашёл ещё пару мест, где они хорошо смотрятся

Versach commented on src/test/java/com/modsensoftware/marketplace/unit/item/ItemDaoTest.java:
```diff
@@ -0,0 +1,263 @@
+        Item result = underTest.get(item.getId());
+        Assertions.assertThat(result).isEqualTo(expected);
+
+        // clean up
```
afterEach не получится использовать? ну или хотябы в метод отдельный вынести кажется местами ты одно и тоже постоянно делаешь и нарушаешь DRY =)

S1mm1ssim replied:
Через AfterEach получится только схему сносить и пересоздавать. Мне кажется, это очень радикально.
Я вынес удаление в методы

Versach commented on src/test/java/com/modsensoftware/marketplace/unit/item/ItemServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+        // then
+        Assertions.assertThatThrownBy(() -> underTest.updateItem(id, updatedFields))
+                .isInstanceOf(OptimisticLockException.class)
+                .hasMessage("Provided item version does not match with the one in the database");
```
Вот видишь какая ерунда получается, а было б в константе - не надо было б строки повтораять

S1mm1ssim replied:
Сделал такой вариант с hardcoded сообщениями в константах. Эти значения сетаются с помощью ReflectionUtils в сервисы. Если так подумать, то нам без разницы какое сообщение там будет в действительности. Ведь валидируем  то, что сообщение ровно такое, какое мы ожидаем, а не само его значение, верно?

Versach replied:
ReflectionUtils совсем необязательно юзать, но в юнитах так лучше наверное будет да

Versach commented on src/test/java/com/modsensoftware/marketplace/unit/item/ItemServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+        ItemDto updatedFields = new ItemDto();
+        updatedFields.setVersion(version);
+        updatedFields.setDescription("description");
+        Item item = new Item();
```
где возможно - создание тестовых сущностей стоит так же выносить в методы и переисопльзовать их

S1mm1ssim replied:
Вроде бы везде поправил

Versach reviewed (commented):
много проблем со строковыми константами и повторений за самим собой

S1mm1ssim committed:
Remove log.isDebugEnabled() in project

S1mm1ssim committed:
Move reusable DEFAULT_PAGE_NUMBER constant to Constants.java

S1mm1ssim committed:
Refactor SessionFactoryConfig

- Remove DependsOn hibernateProperties bean because it is not needed for init
- Move exception message to constant

S1mm1ssim committed:
Move Constants.java to package

S1mm1ssim committed:
Move filter names to Constants.java and import them to services and DAOs

S1mm1ssim committed:
Move exception messages to constants

S1mm1ssim committed:
Move exception messages to constants and remove magic number

S1mm1ssim committed:
Move exception messages to properties file

S1mm1ssim committed:
Make Constants.java final

S1mm1ssim committed:
Move entity field name constants from DAO to entity classes

S1mm1ssim committed:
Remove .* imports

S1mm1ssim committed:
Move default test run properties to constants

S1mm1ssim committed:
Improve readability of CategoryDaoTest tests

S1mm1ssim committed:
Refactor CompanyDao unit tests

- Remove code duplication
- Replace setter object creation with constructor

S1mm1ssim committed:
Refactor ItemDao unit tests

- Remove code duplication
- Validate exception message equal to property
- Rename some variables

S1mm1ssim committed:
Refactor UserDao unit tests

- Remove code duplication
- Validate exception message equal to property
- Rename several variables
- Add some static imports

S1mm1ssim committed:
Update service tests to set exception messages in services with reflect utils

S1mm1ssim committed:
Refactor CompanyDao tests

- Move test entity creation to methods

S1mm1ssim committed:
Add missing logs

S1mm1ssim committed:
Implement parametrized tests

S1mm1ssim committed:
Refactor integration tests

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/CompanyController.java:
```diff
@@ -30,45 +35,39 @@ public class CompanyController {
-        return companyService.getAllCompanies();
+    public List<Company> getAllCompanies(
+            @RequestParam(name = "page", defaultValue = DEFAULT_PAGE_NUMBER) int pageNumber,
+            @RequestParam(name = EMAIL_FILTER_NAME, required = false) String email,
```
Валидацию не будешь добавлять?

S1mm1ssim replied:
Извини, немного не понимаю. А о какой валидации ты говоришь?

S1mm1ssim replied:
Добавил в этой ветке

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/ItemController.java:
```diff
@@ -31,44 +33,35 @@ public class ItemController {
-        }
-        return itemService.getAllItems();
+    public List<Item> getAllItems(@RequestParam(name = "page",
+            defaultValue = Constants.DEFAULT_PAGE_NUMBER) int pageNumber) {
```
Если импортнуть константу/ы static импортом будет читабельнее

S1mm1ssim replied:
Готово

Versach replied:
? на вид не помнялось ничего

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/PositionController.java:
```diff
@@ -30,44 +32,35 @@ public class PositionController {
-            log.debug("Fetching all positions");
-        }
-        return positionService.getAllPositions();
+    public List<Position> getAllPositions(@RequestParam(name = "page",
```
page тоже просится в константы все ж как никак на уровне твоей апи думаю это стандартное название параметра

S1mm1ssim replied:
Готово

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/PositionController.java:
```diff
@@ -30,44 +32,35 @@ public class PositionController {
     }
 
     @GetMapping(value = "/{id}", produces = {"application/json"})
     public Position getPositionById(@PathVariable(name = "id") Long id) {
```
как и id

S1mm1ssim replied:
Точно, забыл про него. Сделал

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/UserController.java:
```diff
@@ -31,44 +38,40 @@ public class UserController {
-        }
-        return userService.getAllUsers();
+    public List<User> getAllUsers(
+            @RequestParam(name = "page", defaultValue = DEFAULT_PAGE_NUMBER) int pageNumber,
```
Обрати внимание, ты все вынес в константы а page нет, чем тебе page не угодил?)

S1mm1ssim replied:
Исправил)

Versach commented on src/test/java/com/modsensoftware/marketplace/unit/category/CategoryDaoTest.java:
```diff
@@ -0,0 +1,199 @@
+        // then
+        Assertions.assertThatThrownBy(() -> underTest.get(nonexistentId))
+                .isInstanceOf(EntityNotFoundException.class)
+                .hasMessage(format(categoryNotFoundMessage, nonexistentId));
```
видишь как хорошо когда тсроки в перменных а так бы пришлось текст сюда валить

S1mm1ssim replied:
Согласен, буду так всегда делать теперь

Versach commented on src/test/java/com/modsensoftware/marketplace/unit/category/CategoryMapperTest.java:
```diff
@@ -0,0 +1,82 @@
+
+    private final CategoryMapper underTest = new CategoryMapperImpl();
+
+    @ValueSource(longs = {1})
```
кажется подразумевалось больше одного теста иначе зачем?

S1mm1ssim replied:
Там ещё NullSource был и получается, что 2 теста было, но хорошо, добавил ещё.

Versach commented on src/test/java/com/modsensoftware/marketplace/unit/category/CategoryMapperTest.java:
```diff
@@ -0,0 +1,82 @@
+        Assertions.assertThat(actual.getParent()).isEqualTo(expected.getParent());
+    }
+
+    @ValueSource(booleans = {false})
```
тот же вопрос, зачем valueSource для одного кейса. Надо добавить других

S1mm1ssim replied:
Тут ещё помимо false була тест вызывается с null значением один раз. У этих двух случаев маппинга одинаковый результат. Но если ещё true добавить, то тест упадет, потому что если приходит true, то совершенно другой результат маппинга, и тест на такой граничный случай у меня есть shouldMapDtoToCategoryWithProvidedParentIdValue(null)

Versach replied:
ок, в valuesource нельзя null передать? ( не помню)

S1mm1ssim replied:
Нет, нельзя, потому что как я понимаю в ValueSource примитивы только можно передавать

Versach replied:
Ок

Versach commented on src/test/java/com/modsensoftware/marketplace/unit/position/PositionServiceTest.java:
```diff
@@ -0,0 +1,76 @@
+        long version = 1L;
+        long differentVersion = 2L;
+        double amount = 10d;
+        PositionDto updatedFields = new PositionDto(null, null, null, amount, version);
```
Может стоить заюзать builder?

S1mm1ssim replied:
Да, стало лучше выглядеть. Добавил

Versach reviewed (commented):
По большей части все ок, правь мелочи и можно мержить будет

S1mm1ssim committed:
Move "page" filter name to Constants.java

S1mm1ssim committed:
Move id path variable name to Constants.java

S1mm1ssim committed:
Add border values to test

S1mm1ssim committed:
Add lombok Builder annotations to Item, Position and their DTOs. Update tests

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/PositionController.java:
```diff
@@ -30,44 +35,36 @@ public class PositionController {
-        }
-        return positionService.getAllPositions();
+    public List<Position> getAllPositions(
+            @RequestParam(name = PAGE_FILTER_NAME, defaultValue = DEFAULT_PAGE_NUMBER) int pageNumber) {
```
pageNumber тоже просится валидация

S1mm1ssim replied:
Добавил

Versach commented on src/main/java/com/modsensoftware/marketplace/controller/PositionController.java:
```diff
@@ -30,44 +35,36 @@ public class PositionController {
 
     @ResponseStatus(HttpStatus.CREATED)
     @PostMapping
     public void createPosition(@RequestBody PositionDto positionDto) {
```
Внутрь positionDto если нет тоже стоит валидации добавить

S1mm1ssim replied:
Добавил

S1mm1ssim committed:
Add parameter validation
